171648169 Feature of signing in a user

Irutingabo opened the pull request:
#### Description
> This PR adds the sign-in feature to the application
#### Type of change

Please select the relevant option

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Others (cosmetics, styling, improvements)
- [ ] This change requires a documentation update

#### How Has This Been Tested?

- [x] Unit
- [ ] Integration
- [ ] End-to-end
- [ ] No testing required

#### How to Test
> Use postman to test the sign-in feature,
Give instructions on how reviewers can verify your work

#### Any background context you want to add (Optional)
> N/A
Screenshots or GIFs
> Mockup
https://www.figma.com/file/2KDdSdzTye8EG7l0Wxl36D/Barefoot-dev?node-id=1%3A3

>Implementation plan
https://docs.google.com/document/d/134cKdaAu0KEDQYztmDO0ljIuTO16WvA_eeDhhUhMLPE/edit?usp=sharing
#### STORY ID

[[Finishes #171648169]

YankyJayChris commented on server/controllers/authController.js:
```diff
@@ -0,0 +1,58 @@
+            }
+          });
+        });
+      },
+    )(req, res);
+  }
```
what is the use of this ?

Irutingabo replied:
Every time an API request is made, the request (req) passes through the authentication strategy used by passport, and if successfully authenticated, then the passport binds the details of the user to the request(req).

YankyJayChris commented on server/controllers/authController.js:
```diff
@@ -0,0 +1,58 @@
+     * @returns {object} The status and some data of the user.
+     */
+  static signIn(req, res) {
+    passport.authenticate(
+      'local',
+      { session: false },
+      (error, user) => {
+        if (error || !user) {
+          res.status(404).json({ error });
+        }
+
+        const { id, email, lastname } = user;
+        /** This is what goes in our JWT */
+        const payload = {
+          id,
+          expires: Date.now() + parseInt(JWT_EXPIRES_MS, 4),
+        };
+
+        /** assigns payload to req.user */
+        req.login(payload, { session: false }, (err) => {
+          if (err) {
+            res.status(400).send({ err });
+          }
```
i think this good but it can be better if you add them in middle ware so that everyone can have access on the signed in user using that middle ware.

Irutingabo replied:
@YankyJayChris  thanks for your feedback,
If I got it right, You want me to separate a common functionality like
Generating token, I can do that, but this is only signing in functionality, I don't think you will need to access inside functions. if you do please feel free to use this static method `signIn`
otherwise, if you see any other function that you need to use, please suggest it!

placideirandora commented on .eslintrc:
```diff
@@ -14,12 +14,16 @@
         "import/no-unresolved": [2, {
             "commonjs": true
         }],
+        "no-plusplus" : "off",
```
Consider turning this rule back on and use this format. A = A + 1 instead.

Irutingabo replied:
It's done
![image](https://user-images.githubusercontent.com/11549858/78555414-440f9000-780d-11ea-8d4f-ed6a4dee0fa5.png)

![image](https://user-images.githubusercontent.com/11549858/78555652-cbf59a00-780d-11ea-82de-3482cddacc6b.png)

placideirandora commented on server/config/passport.js:
```diff
@@ -0,0 +1,26 @@
+  passwordField: 'password',
+}, async (email, password, done) => {
+  try {
+    const userDocument = await AuthHelpers.userExists('email', email);
```
Consider renaming the variable from **userDocument** to something like **user** or **userExists**

Irutingabo replied:
Feedback implemented... renamed to user

placideirandora commented on server/middlewares/validateAuth.js:
```diff
@@ -0,0 +1,30 @@
+   * @param {object} next pass to next method.
+   * @returns {object} The error and some error messages .
+   */
+  static validateSignin(req, res, next) {
```
Consider capitalizing the **i** as well to become **validateSignIn**.

Irutingabo replied:
Feedback implemented!

placideirandora commented on server/middlewares/validateAuth.js:
```diff
@@ -0,0 +1,30 @@
+      next();
+    } else {
+      const { details } = error;
+      const message = details.map((i) => i.message).join(',');
```
Consider resolving this issue.

placideirandora commented on server/routes/auth.js:
```diff
@@ -0,0 +1,13 @@
+import AuthController from '../controllers/authController';
+import bnLog from '../helpers/log.util';
+import validateAuth from '../middlewares/validateAuth';
+// eslint-disable-next-line no-unused-vars
```
Resolve this issue as well. You should it disable eslint rules. Refer to their documentation on how to fix its issues.

placideirandora commented on server/routes/auth.js:
```diff
@@ -0,0 +1,13 @@
+const router = Router();
+
+router
+  .post('/signin', validateAuth.validateSignin, bnLog(AuthController.signIn));
```
Capitalize the last i as well: **validateSignIn**.

Irutingabo replied:
Feedback implemented... Well renamed!

placideirandora commented on server/schemas/authSchema.js:
```diff
@@ -0,0 +1,8 @@
+/* eslint-disable import/prefer-default-export */
```
Resolve this issue as well.

Irutingabo replied:
Feedback implemented...

placideirandora commented on server/tests/auth-signin.test.js:
```diff
@@ -0,0 +1,68 @@
+import { describe, it } from 'mocha';
+import chaiHttp from 'chai-http';
+import app from '../index';
+import 'dotenv';
```
Resolve this issue

placideirandora commented on server/tests/auth-signin.test.js:
```diff
@@ -0,0 +1,59 @@
+      .post('/api/v1/signin')
+      .send(users[1])
+      .end((err, res) => {
+        expect(res).to.have.status(422);
```
You should test all response elements such as the returned message as well.

Irutingabo replied:
Feedback implemented... tested all response

placideirandora commented on server/tests/auth-signin.test.js:
```diff
@@ -0,0 +1,59 @@
+      .post('/api/v1/signin')
+      .send(users[2])
+      .end((err, res) => {
+        expect(res).to.have.status(404);
```
You should test all response elements such as the returned message as well.

Irutingabo replied:
Feedback implemented... tested all response

placideirandora commented on server/tests/auth-signin.test.js:
```diff
@@ -0,0 +1,59 @@
+      .post('/api/v1/signin')
+      .send(users[3])
+      .end((err, res) => {
+        expect(res).to.have.status(422);
```
You should test all response elements such as the returned message as well.

Irutingabo replied:
Feedback implemented... tested all response

placideirandora commented on server/tests/auth-signin.test.js:
```diff
@@ -0,0 +1,59 @@
+      .post('/api/v1/signin')
+      .send(users[4])
+      .end((err, res) => {
+        expect(res).to.have.status(422);
```
The same applies here

Irutingabo replied:
Feedback implemented... tested all response

placideirandora commented on server/tests/auth-signin.test.js:
```diff
@@ -0,0 +1,59 @@
+      .post('/api/v1/signin')
+      .send(users[5])
+      .end((err, res) => {
+        expect(res).to.have.status(404);
```
The same applies here

Irutingabo replied:
Feedback implemented... tested all response

Irutingabo commented on .eslintrc:
```diff
@@ -14,12 +14,16 @@
         "import/no-unresolved": [2, {
             "commonjs": true
         }],
+        "no-plusplus" : "error",
```
Resolved!
https://github.com/STACK-UP-3/champs-backend/pull/12#discussion_r403117578

devel605 committed:
171648169-story(sign-in): Adding Sign in feature

- Prepared Mockup
- Added validation to inputs
- Verify User existance in Database
- Encrypted Password
- Added Passport and JWT to generete token for future authentication
- Updated swagger
- Tested

[Finishes #171648169]
